### PR TITLE
Master website portal chatter pgu

### PR DESCRIPTION
**Pull request by pgu-odoo**
Description of the issue/feature this PR addresses:
task id -
https://www.odoo.com/web#id=37264&action=327&model=project.task&view_type=form&menu_id=
Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/portal/controllers/mail.py [hunk 1]
```diff
@@ -66,12 +72,12 @@ def _message_post_helper(res_model='', res_id=None, message='', token='', nosubs
...
     @http.route(['/mail/chatter_post'], type='http', methods=['POST'], auth='public', website=True)
-    def portal_chatter_post(self, res_model, res_id, message, **kw):
+    def portal_chatter_post(self, res_model, res_id, message, attachments, **kw):
         url = request.httprequest.referrer
-        if message:
+        if message or attachments:
             # message is received in plaintext and saved in html
             message = plaintext2html(message)
-            _message_post_helper(res_model, int(res_id), message, **kw)
+            _message_post_helper(res_model, int(res_id), message, attachments, **kw)
```
[the remark this answers is not in the document]
Comment (Xavier-Do) [reply addons/portal/controllers/mail.py]: I meant something like:

_message_post_helper(res_model=res_model, res_id=int(res_id), message=message, attachments=attachments, **kw)

File: addons/portal/controllers/mail.py [hunk 2]
```diff
@@ -58,6 +60,7 @@ def _message_post_helper(res_model='', res_id=None, message='', token='', nosubs
     kw.pop('attachment_ids', None)
```
Comment (Xavier-Do) [inline addons/portal/controllers/mail.py]: this line looks useless

File: odoo/addons/base/models/ir_attachment.py [hunk 3]
```diff
@@ -50,7 +50,7 @@ def _compute_res_name(self):
     def _compute_res_model_name(self):
         for record in self:
             if record.res_model:
-                model = self.env['ir.model'].search([('model', '=', record.res_model)], limit=1)
+                model = self.env['ir.model'].sudo().search([('model', '=', record.res_model)], limit=1)
```
Comment (Xavier-Do) [inline odoo/addons/base/models/ir_attachment.py]: It isn't something to change now but with  PR https://github.com/odoo/odoo/pull/28563, we will be able to remove this sudo.

File: addons/portal/controllers/mail.py [hunk 4]
```diff
@@ -144,6 +157,102 @@ def portal_message_fetch(self, res_model, res_id, domain=False, limit=10, offset
...
+                    'filename': filename,
+                    'mimetype': ufile.content_type,
+                    'id': attachment.id,
+                    'size': attachment.file_size
+                })
+
+        return out % (json.dumps(callback), json.dumps(args))
+
+    @http.route('/portal/content/<int:attachment_id>', type='http', auth="public")
+    def download_attachment(self, attachment_id, model, id, access_token=False, field="datas", filename_field='name', download=None, **kw):
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/portal/controllers/mail.py]: I don't think `field` and `filename_field` are necessary here, they would allow the user to read any field of the attachment for no reason.

File: addons/portal/static/src/js/portal_chatter.js [hunk 5]
```diff
@@ -34,6 +38,20 @@ var PortalComposer = publicWidget.Widget.extend({
...
         });
+        this.set('attachmentIDs', []);
+        this.fileuploadID = _.uniqueId('o_chat_fileupload');
+    },
+    /**
+     * @override
+     */
+    start: function () {
+        //portal chatter
+        this.$attachmentButton = this.$(".o_portal_chatter_attachment_btn");
```
Comment (seb-odoo) [inline addons/portal/static/src/js/portal_chatter.js]: Single quote here and below.

File: addons/portal/static/src/js/portal_chatter.js [hunk 6]
```diff
@@ -46,6 +64,128 @@ var PortalComposer = publicWidget.Widget.extend({
...
+    /**
+     * @private
+     */
+    _onAttachmentChange: function (ev) {
+        var files = ev.target.files,
+            attachments = this.get('attachmentIDs');
+
+        if (attachments.length + files.length > 10) {
+            this.display_alert(_t("Oops! You can not upload more than 10 files."));
+        }
```
Comment (seb-odoo) [inline addons/portal/static/src/js/portal_chatter.js]: Linter: else on same line.

File: addons/portal/static/src/js/portal_chatter.js [hunk 7]
```diff
@@ -46,6 +64,128 @@ var PortalComposer = publicWidget.Widget.extend({
...
+        else if (!_.isEmpty(_.filter(files, function (file) {return (file.size / 1024 / 1024) > 25;}))) {
+            this.display_alert(_t("Oops! You can not upload a file larger than 25MB."));
+        }
```
Comment (seb-odoo) [inline addons/portal/static/src/js/portal_chatter.js]: Linter: else on same line.

File: addons/portal/static/src/js/portal_chatter.js [hunk 8]
```diff
@@ -46,6 +64,128 @@ var PortalComposer = publicWidget.Widget.extend({
...
+     * @private
+     */
+    _onAttachmentChange: function (ev) {
+        var files = ev.target.files,
+            attachments = this.get('attachmentIDs');
+
+        if (attachments.length + files.length > 10) {
+            this.display_alert(_t("Oops! You can not upload more than 10 files."));
+        }
+        else if (!_.isEmpty(_.filter(files, function (file) {return (file.size / 1024 / 1024) > 25;}))) {
```
Comment (seb-odoo) [inline addons/portal/static/src/js/portal_chatter.js]: Linter: space after `{` and before `}` and statement in a new line

File: addons/portal/static/src/js/portal_chatter.js [hunk 9]
```diff
@@ -46,6 +64,128 @@ var PortalComposer = publicWidget.Widget.extend({
...
+            this.display_alert(_t("Oops! You can not upload a file larger than 25MB."));
+        }
+        else {
+            _.each(files, function (file) {
+                var duplicate_attachment = _.findWhere(attachments, {name: file.name});
```
Comment (seb-odoo) [inline addons/portal/static/src/js/portal_chatter.js]: Linter: use camel case

File: addons/portal/static/src/js/portal_chatter.js [hunk 10]
```diff
@@ -46,6 +64,128 @@ var PortalComposer = publicWidget.Widget.extend({
...
+     * @private
+     */
+    _onAttachmentLoaded: function (ev) {
+        var attachments = this.get('attachmentIDs'),
+            files = Array.prototype.slice.call(arguments, 1);
+
+        _.each(files, function (file) {
+            if (file.error || !file.id) {
+                this.display_alert(file.error);
+                attachments = _.filter(attachments, function (attachment) { return !attachment.upload; });
```
Comment (seb-odoo) [inline addons/portal/static/src/js/portal_chatter.js]: Linter: statement in new line

File: addons/portal/static/src/scss/portal.scss [hunk 11]
```diff
@@ -486,7 +486,7 @@ form label {
 
     .o_portal_page_size {
         width: 100%;
-        max-width: 210mm; // A4 page's width
+        max-width: 235mm;
```
Comment (seb-odoo) [inline addons/portal/static/src/scss/portal.scss]: Why this change?

File: addons/portal/controllers/mail.py [hunk 12]
```diff
@@ -144,6 +157,102 @@ def portal_message_fetch(self, res_model, res_id, domain=False, limit=10, offset
...
+            try:
+                document.check_access_rights('read')
+                document.check_access_rule('read')
+            except AccessError:
+                if not access_token or not _check_special_access(model, int(id), token=access_token):
+                    raise Forbidden()
+        except (AccessError, MissingError):
+            raise Forbidden()
+
+        record = document_sudo.website_message_ids.mapped('attachment_ids').filtered(lambda x: x.id == int(attachment_id))
```
Comment (seb-odoo) [inline addons/portal/controllers/mail.py]: Do you really need to filter them like that? For performance it would be better to make a `search` on attachment for the given `id`, `model` and `attachment_id`. Or make a `browse` with `attachment_id` and then make sure `model` and `id` match.

Comment (seb-odoo) [reply addons/portal/controllers/mail.py]: This might be to filter attachments put on notes? Then a check can also be put on `message_type` of `comment`.

Comment (seb-odoo) [reply addons/portal/controllers/mail.py]: No it doesn't work, so that's a problem to solve too: the public user can currently see the attachments put on notes by changing the `attachment_id` in the url.

File: addons/portal/static/src/scss/portal.scss (hunk 11 above)
Comment (pgu-odoo) [reply addons/portal/static/src/scss/portal.scss]: Spacing issue,mention in pad  - https://drive.google.com/file/d/1ZNXotoO9wjZ5iE9SF_z6Hj0ScNBQRD-4/view?usp=drivesdk

File: addons/portal/controllers/mail.py (hunk 12 above)
Comment (pgu-odoo) [reply addons/portal/controllers/mail.py]: field website_message_ids already have domain for 'comment' type messgae.
https://github.com/odoo/odoo/blob/master/addons/portal/models/mail_thread.py#L16

File: addons/portal/controllers/mail.py [hunk 13]
```diff
@@ -144,6 +160,102 @@ def portal_message_fetch(self, res_model, res_id, domain=False, limit=10, offset
...
+            if request.httprequest.user_agent.browser == 'safari':
+                # Safari sends NFD UTF-8 (where é is composed by 'e' and [accent])
+                # we need to send it the same stuff, otherwise it'll fail
+                filename = unicodedata.normalize('NFD', ufile.filename)
+            try:
+                attachment = AttachmentSudo.create({
+                    'name': filename,
+                    'datas': base64.encodebytes(data),
+                    'datas_fname': filename,
+                    'res_model': model,
```
Comment (Xavier-Do) [inline addons/portal/controllers/mail.py]: When doing a two step upload (upload attachments then giving ids in message post) the attachment should be added on mail.compose.message with id 0. Message post will add it to the record automatically. This will avoid to add attachment to record if user cancel message writing.

File: addons/portal/controllers/mail.py [hunk 14]
```diff
@@ -91,12 +103,16 @@ def _message_post_helper(res_model, res_id, message, token='', nosubscribe=True,
 class PortalChatter(http.Controller):
 
     @http.route(['/mail/chatter_post'], type='http', methods=['POST'], auth='public', website=True)
-    def portal_chatter_post(self, res_model, res_id, message, **kw):
+    def portal_chatter_post(self, res_model, res_id, message, attachment_ids=None, **kw):
         url = request.httprequest.referrer
-        if message:
+        if attachment_ids:
```
Comment (Xavier-Do) [inline addons/portal/controllers/mail.py]: Giving some correct access token the attachment_ids will be read as sudo in message_post, we need to be sure that the attachments are accessible before this step. Unfortunately it is difficult to know. If we ignore the comment below about attachments model and res_id at create, we may want check that attachment res_ids and res_model are matching the record, but it could come from a private note, so we may want to check that the attachment is not linked to any message yet. 
If we link the attachment to mail.compose.message, we could check that model of all attachment_ids is mail.compose.message and res_id is 0. 

So a big security issue here I think, I need to check this in details.

File: addons/portal/controllers/mail.py [hunk 15]
```diff
@@ -91,12 +103,16 @@ def _message_post_helper(res_model, res_id, message, token='', nosubscribe=True,
...
+            attachment_ids = literal_eval(attachment_ids)
+            if isinstance(attachment_ids, int):
+                attachment_ids = [attachment_ids]
+        if message or attachment_ids:
             # message is received in plaintext and saved in html
             message = plaintext2html(message)
-            _message_post_helper(res_model, int(res_id), message, **kw)
+            _message_post_helper(res_model=res_model, res_id=int(res_id), message=message, attachment_ids=attachment_ids, **kw)
```
Comment (Xavier-Do) [inline addons/portal/controllers/mail.py]: Note, not really linked to pr but we need to filter kw here to avoid injections in message post.

File: addons/portal/controllers/mail.py [hunk 16]
```diff
@@ -144,6 +160,102 @@ def portal_message_fetch(self, res_model, res_id, domain=False, limit=10, offset
             'message_count': Message.search_count(domain)
         }
 
+    @http.route('/portal/binary/upload_attachment', type='http', auth="public")
+    def upload_attachment(self, callback, model, id, ufile, **kwargs):
```
[the remark this answers is not in the document]
Comment (mba-odoo) [reply addons/portal/controllers/mail.py]: @Xavier-Do ,  I am not in favor to add public road for upload attachments. 
but as per spec when we share a particular record using token we should allow to upload the attachment on that record. 
so we used parent record token to upload an attachment on a particular record. I do not think there is any other way to upload without a public route.
we should restrict the number of attachment on a particular record by the public as well as by the portal user. what do you think?

Thanks

[the remark this answers is not in the document]
Comment (Xavier-Do) [reply addons/portal/controllers/mail.py]: In this case if a record is editable by a public user, no token is required, this is what is a problem for me.  I think that in odoo we don't have any model inheriting from mail_thread with such an access right, but i feel like it is still a good idea to check that user is not in group public if no access token is given.

File: addons/portal/controllers/mail.py (hunk 15 above)
Comment (mba-odoo) [reply addons/portal/controllers/mail.py]: @Xavier-Do ,  we already filter some kw in `_message_post_helper`. Thanks

File: addons/portal/controllers/mail.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (mba-odoo) [reply addons/portal/controllers/mail.py]: @Xavier-Do , done

File: addons/portal/controllers/mail.py (hunk 14 above)
Comment (mba-odoo) [reply addons/portal/controllers/mail.py]: @Xavier-Do , we can not use the model 'mail.compose.message' and res_id as 0 because when attachments are added by the internal user. it adds the record's model and res_id on attachment and we want to allow it to download by the public user. (For Ex: while sending a quotation, we attach the quotation pdf and should be accessible by the public user when he access the quotation ) 

as per your first point, we are using the res_id and model and only using those message's attachment which is linked with `website_message_ids` and also checking for the model and res_id at download.  Thanks

**Comment by seb-odoo** [comment]
This will be merged in #34360 (invoice style), #34384 (fix download) and #4526 (new upload)
